addons/better-terrain/editor/TileView.gd

```diff
@@ -263,77 +396,250 @@ func _draw() -> void:
-			(tile.valid and tile.get("peering") != highlighted_tile_part.get("peering")):
+			(tile.valid and tile.get("peering") != highlighted_tile_part.get("peering")) or\
+			event.button_mask & MOUSE_BUTTON_LEFT and paint_action == PaintAction.SELECT:
 			queue_redraw()
 		highlighted_tile_part = tile
 	
-	var clicked : bool = event is InputEventMouseButton and event.pressed
-	if paint >= 0 and clicked:
-		paint_action = PaintAction.NO_ACTION
-		if !highlighted_tile_part.valid:
-			return
-		
-		match [paint_mode, event.button_index]:
-			[PaintMode.PAINT_TYPE, MOUSE_BUTTON_LEFT]: paint_action = PaintAction.DRAW_TYPE
-			[PaintMode.PAINT_TYPE, MOUSE_BUTTON_RIGHT]: paint_action = PaintAction.ERASE_TYPE
-			[PaintMode.PAINT_PEERING, MOUSE_BUTTON_LEFT]: paint_action = PaintAction.DRAW_PEERING
-			[PaintMode.PAINT_PEERING, MOUSE_BUTTON_RIGHT]: paint_action = PaintAction.ERASE_PEERING
+	var clicked : bool = event is InputEventMouseButton and (event.pressed and (event.button_index == MOUSE_BUTTON_LEFT or event.button_index == MOUSE_BUTTON_RIGHT))
+	if clicked:
+		initial_click = current_position
+		selection_start = Vector2i(-1,-1)
+	if released:
+		selection_rect = Rect2i(0,0,0,0)
+		queue_redraw()
 	
-	if (clicked or event is InputEventMouseMotion) and paint_action != PaintAction.NO_ACTION:
-		if !highlighted_tile_part.valid:
-			return
+	if paint_action == PaintAction.PASTE:
+		if event is InputEventMouseMotion:
+			queue_redraw()
 		
-		if paint_action == PaintAction.DRAW_TYPE or paint_action == PaintAction.ERASE_TYPE:
-			var type := BetterTerrain.get_tile_terrain_type(highlighted_tile_part.data)
-			var goal := paint if paint_action == PaintAction.DRAW_TYPE else -1
-			if type != goal:
-				undo_manager.create_action("Set tile terrain type", UndoRedo.MERGE_DISABLE, tileset)
-				undo_manager.add_do_method(BetterTerrain, &"set_tile_terrain_type", tileset, highlighted_tile_part.data, goal)
-				undo_manager.add_do_method(self, &"queue_redraw")
-				if goal == -1:
-					terrain_undo.create_peering_restore_point_tile(
-						undo_manager,
-						tileset,
-						highlighted_tile_part.source_id,
-						highlighted_tile_part.coord,
-						highlighted_tile_part.alternate
-					)
-				else:
-					undo_manager.add_undo_method(BetterTerrain, &"set_tile_terrain_type", tileset, highlighted_tile_part.data, type)
-				undo_manager.add_undo_method(self, &"queue_redraw")
-				undo_manager.commit_action()
-		elif paint_action == PaintAction.DRAW_PEERING:
-			if highlighted_tile_part.has("peering"):
-				if !(paint in BetterTerrain.tile_peering_types(highlighted_tile_part.data, highlighted_tile_part.peering)):
-					undo_manager.create_action("Set tile terrain peering type", UndoRedo.MERGE_DISABLE, tileset)
-					undo_manager.add_do_method(BetterTerrain, &"add_tile_peering_type", tileset, highlighted_tile_part.data, highlighted_tile_part.peering, paint)
+		if clicked:
+			if event.button_index == MOUSE_BUTTON_LEFT:
```

Comment:
There's a lot of indentation here, though it seems some may be unavoidable. I'd combine these if statements into
`if event.button_index == MOUSE_BUTTON_LEFT and taged_paste_tile_states.size() > 0:`